### ISSUE TITLE
Is this a fatal error in ssl_write_real()? I'm not sure

### ISSUE THREAD

**Issue by dhhome2006** [ARMmbed/mbedtls#1707]
### Description
- Type    : Bug
- Priority: Major

## Bug
**OS**  
windows xp sp3, ubuntu

**mbed TLS build:**  
Version : mbedtls-2.9.0-apache.tgz
Compiler: vs2010, gcc4.9

{{{
Maybe there is a bug,

package : mbedtls-2.9.0-apache.tgz
file    : ssl_tls.c
line    : 7249

current code:
```
    if( ssl->out_left != 0 )
    {
        if( ( ret = mbedtls_ssl_flush_output( ssl ) ) != 0 )
        {
            MBEDTLS_SSL_DEBUG_RET( 1, "mbedtls_ssl_flush_output", ret );
            return( ret );
        }
    }
    else
    {
        ssl->out_msglen  = len;
        ssl->out_msgtype = MBEDTLS_SSL_MSG_APPLICATION_DATA;
        memcpy( ssl->out_msg, buf, len );

        if( ( ret = mbedtls_ssl_write_record( ssl ) ) != 0 )
        {
            MBEDTLS_SSL_DEBUG_RET( 1, "mbedtls_ssl_write_record", ret );
            return( ret );
        }
    }
    return( (int) len );
```

fixed code:
```
    if( ssl->out_left != 0 )
    {
        if( ( ret = mbedtls_ssl_flush_output( ssl ) ) != 0 )
        {
            MBEDTLS_SSL_DEBUG_RET( 1, "mbedtls_ssl_flush_output", ret );
            return( ret );
        }
    }

    ssl->out_msglen  = len;
    ssl->out_msgtype = MBEDTLS_SSL_MSG_APPLICATION_DATA;
    memcpy( ssl->out_msg, buf, len );

    if( ( ret = mbedtls_ssl_write_record( ssl ) ) != 0 )
    {
        MBEDTLS_SSL_DEBUG_RET( 1, "mbedtls_ssl_write_record", ret );
        return( ret );
    }
    return( (int) len );
```

Eric Tung, dhhome2006@gmail.com
}}}

**Comment by sbutcher-arm** [comment]
I'm not sure this is a real bug, but we'll look into it, and maybe we need to improve documentation.

cc: @RonEld

**Comment by ciarmcom** [comment]
ARM Internal Ref: IOTSSL-2357

**Comment by RonEld** [comment]
@dhhome2006 Thank you for raising this issue!
I believe you may have a point.
In current implementation,  in rare cases where the output buffer already contains data to send by the time the application calls `mbedtls_ssl_write()`, then this implementation flushes the output buffer, and assuming the flushing succeeded, returns the length of the message (or the max fragment length if it's shorter), as the number of bytes sent from the input message. but the message wasn't sent at all.

In your suggestion, `mbedtls_write_real()` first flushes the output buffer, and then sends the input data, and returns the actual bytes sent from the input message.

I wouldn't say it's a fatal error, but I do believe this is a bug.
Would you be willing to submit a PR resolving this issue?

**Comment by dhhome2006** [comment]
You are right, this situation is very rare, basically it only happens in the server with the large data traffic. For example, we are using the video forwarding service.
I'll make a PR later, thank you!

**Comment by RonEld** [comment]
@dhhome2006 After looking some more, I believe this issue is already addressed and fixed in https://github.com/ARMmbed/mbedtls/issues/839

**Comment by RonEld** [comment]
I take it back, I don't think it fixes the issue

**Comment by dhhome2006** [comment]
@RonEld  I looked at the fix for this problem by @davidwu2000, but I think the semantics of mbedtls_ssl_flush_output return 0 are different from the semantics of ssl_write_real, although there is nothing wrong with this fix!

**Comment by RonEld** [comment]
@dhhome2006 In this case, We will close this issue, and your PR, as it is being addressed by #1006 
Please note that the fix is documentation change, and you should verify your application is according to the documentation.
